hawkular-datamining-api/src/main/java/org/hawkular/datamining/api/base/InMemorySubscriptionManager.java

```diff
@@ -74,8 +67,9 @@ public void setPredictionListener(PredictionListener predictionListener) {
     }
 
     @Override
-    public void subscribe(Subscription subscription) {
+    public synchronized void subscribe(Subscription subscription) {
```

Comment:
I'd maybe synchronize on the `subscriptions` rather than on `this`. Synchronizing on `this` opens a door to a potential deadlock because other code can also synchronize on it. If you synchronize on `subscriptions` you can be sure that the lock is private to your class and that no other lock can exist than the one in `unsubscribe`.